README.md
# Tasmota for Tilt Blinds (Customized)

Tasmota firmware that has been customized for use as a Blinds Controller using an ESP8266 (specifically an ESP12) like the one found at [MK Smarthouse Blinds Controller with MG995 Servo](https://www.mksmarthouse.com/blinds-control). It used to control a "Servo" motor, such as an MG995/996, which is used to control the tilt angle of blind slats and does so in a slow motion manner.  Up to 3 individual blinds can be independently controlled by a single controller.

It has been written for use with an Arduino IDE and tested on board types "Generic" and "NodeMCU". 

The software is based on Sonoff-Tamota version 6.2.1.  It was compiled with "Arduino Core for ESP8266" version 2.3.0.

The software was written in particular for use with Home Assistant's MQTT support (but does not provide MQTT auto-discovery).

See [RELEASENOTES.md](https://github.com/tommyjlong/Blinds-Control/blob/master/RELEASENOTES.md) for release information

# Installation
### Quick Install
Download the released binary from https://github.com/tommyjlong/Blinds-Control/ and flash it to your ESP8266 hardware using your favorite method.

### If You Want to Compile It Yourself on Windows 10
If you want to compile this yourself with the Arduino IDE on Windows 10, here are a few steps to follow:
- Don't yet start Arduino IDE.  If you already have it running, then exit it out.
- Copy the repository's `\lib` files to C:\Users\\_UserName_\Documents\Arduino\ directory. 
  This Assumes Arduino->File->Preferences->Sketch Location set to: C:\Users\\_UserName_\Documents\Arduino.  This also assumes no other interfering libraries are present in this same directory.
- Goto your `sonoff` directory.  Double-Click on the file `sonoff.ino` to startup the Arduino IDE.
- When setting up the Board Manager for esp8266, choose version 2.3.0. 
- Set the board type parameters acording to ESP module being used (See below).
- Click on the "Verify" button (the one with the check mark).  This should compile successfully.
- If successful, you can now flash the compiled binary.

The following devices are supported:
- ESP12<br/>
  Example, see [MK Smarthouse Blinds Controller with MG995 Servo](https://www.mksmarthouse.com/blinds-control)

  You can use the following Arduino IDE Board settings (Choose your own COM port):

  <img src="https://raw.githubusercontent.com/tommyjlong/Blinds-Control/master/media/Board_Generic.jpg" width="250" align="center" />

- Node MCU

  You can use the following Arduino IDE Board settings (Choose your own COM port):

  <img src="https://raw.githubusercontent.com/tommyjlong/Blinds-Control/master/media/Board_NodeMCU.jpg" width="250" align="center" />

## Configuring the Blinds Controller
After flashing the ESP12 and restarting the Blinds Controller, it will come up just like any normal Sonoff-Tasmota device which puts it in WiFi AP mode and will broadcast a sonoff SSID (Ex. SSID= sonoff-4331).  Use a WiFi Client and connect to this AP.
Then use a Web Browser to configure the Blinds-Controller.
### WiFi and Networking Config

- SSID _YourAP-SSID_ 
- Password _Your-AP-password_
- Host _hostname_

Then Hit "Save" which should reboot the ESP.<br/>
Reconnect your WiFi Client to your regular WiFi network then try to connect with your Web Browser to URL http://_hostname_.local.  

Note:  It is recommended that you connect a console to the controller (set Baud Rate to 114200) to see if any errors occur.  After restart, you may see
`WIF: Connect failed with AP incorrect password`.  If so, then goto the console and re-enter your AP's password using the command `password`_Your-AP-password_.

(Optional) Configure Static IP Address:<br/>
The Blinds-Controller by default uses DHCP to get its IP address, etc.  If you would like to set up as static instead, then:<br/>
Goto: Main Menu->Console.  In the "Enter Command" text box, type in the following (hit return each time)
```
ipaddress1 <blinds-control-static-ipaddress>
ipaddress2 <g/w-ipaddress>
ipaddress3 <subnet-mask> Ex. 255.255.255.0
ipaddress4 <dns-ipaddress>
restart 1 
```
The last command restarts the module for these settings to take effect. Note: you may temporarily loose the ability to use http://_hostname_.local If you do, then use http://_blinds-control-static-ipaddress_ instead.

### MQTT
Goto: Main Menu->Configuration->Configure MQTT
- Host: _MQTT Broker/Server IP Address or Name_
- Port: 1883 _This is the default which should work_
- Client: _this Blind-Controller's mqtt client name_

If you require a username and password to authenticate with the MQTT Broker:
- User (DVES_xxxx): _Broker's username_
- Password:  _Broker's password_

Topics by default uses a 3-tier hierarchy with the middle tier generally used as a name for the device with the default being "sonoff".  For example: ex. `stat/sonoff/RESULT`

To change the middle tier name from "sonoff" to something else:
- Topic = _middle-tier-name_   Ex. `my_blinds`

When finished hit "Save"


### Configure the ESP Module
Goto: Main menu->Configuration->Configure Module
- Module Type: The Blinds-Controller firmware retains all the different modules used by the Sonoff-Tasmota v6.2.1 but changes the default to "Generic" which can be used for ESP12 and NodeMCU.  :warning: _Any other module type is to be used at your own risk_.
- GPIO: The Blinds-Controller firmware supports PWM1, PWM2, PWM3 in any combination to control the blinds. Do **NOT** select **PWM1i, PWM2i nor PWM3i**. It has been tested on ESP12 GPIOs 13,14,16. For those using MK Smarthouse Blinds Controller, set GPIO 13 to PWM1.


When finished hit "Save" which will cause it to reboot.

### Miscellaneous ESP Configurations
The following settings are optional:
Goto: Main Menu->Console.  Use the "Enter Command" text box.
- `timezone`_X_ <return> where _X_ is hour from UTC. Ex. New York, X is -4
- `webpassword` _password_ By default no login is needed by the Web broswer, but use this to setup a password (Note: username is "admin").
- `friendlyname` _friendly-name-string_  Use this to change the banner in the Web page.
- `teleperiod` _period_  Periodically Tamota sends out a couple of health status messages. Default is 300s (5 min); Max is 3600s (1hr); 0 to disable


### Checkout and Testing the Blinds Controller

**Checkout**<br/>

The following is optional and can be skipped but otherwise it but can be used as a sanity check.  This will ensure the Blinds-Controller's modification of the Tamota code is set to operate properly. <br/> 
Goto: Main Menu->Console.  In the "Enter Command" text box, type in the following (hit return each time)
- `setoption15`:  This should show `{"SetOption15":"OFF"}` This sets the pwm_control to 0 by default.  0 means "pwm" command is used (See Tasmota Command Documentation)
- `pwmfrequency`: This should show `{"PWMFrequency":50}`.  This is the frequency required by the servos.
- `setoption65`: This should show `{"SetOption65":"ON"}`.  This is a new option that activates the specific usage of PWM to control the blinds (See Note1).
- `setoption4`: This should show `{"SetOption4":"ON"}`.  (See Tasmota Command Documentation for details).
- `setoption66`: This should show `{"SetOption66":"ON"}`. This is a new option that when set to true (and when setoption65 is set to true) will cause the normally published status of a PWM topic to be retained.  This is used for the scenario where Home Assistant reboots, and when it comes back up, it can receive retained MQTT messages. This allows HA to immediately update the last known status of the blind's state and tilt angle.

Note1: The main modification to the Tasmota code is for it to track the last PWM value sent to the servo and changes it incrementally by 1 every time every 50ms (which is every time the Tasmota executive loop runs).  It will continue to increment (or decrement) until it reaches the target PWM value.  This gives it the slow motion effect.

**Test Out and Setup the Servo:** <br/>
Goto: Main Menu->Console and use the "Enter Command" text box. 

The following CLI syntax is to be used: <br/>
     `pwmN` _value_ where N is the PWM channel number 1, 2, or 3 (1 is the default). <br/>
When this command is entered, this should return:  
      `{{"PWM":{"PWM1":<value1>,"PWM2":<value2>, etc.}}` depending on the number of PWM channels configured for the module.

When using for the first time on a servo, the Blinds-Controller does not know what the previous PWM value was (because it never had one) so when entering a PWM value, the servo will run full speed (i.e. not slow-motion).  Use a safe value such as `55`.

The theoretical range for _value_ corresponds to the servo pulse width: 
28 (corresponding to servo angle of 0 degrees) to 123 (180 degrees) with 73 (90 degrees) being middle ways.  However the servo motors in practice are not very precise and vary from motor to motor.  They can even support lower and higher PWM values meaning they can go below 0 degrees and above 180 degrees (but this is not recommended).  Attaching the servo motors to the blinds also adds a variability to the desired tilt angle of the blinds.  

In general these PWM values will have to be determined empirically by playing around with the numbers and watching the tilt postion of the blind's slats, and this will have to be done for each individual blind in a two or three blind set.  One other caution is that when servo motors strain having to use too much force, they will back themselves off at an angle that differs from what the Blinds-Controller tracked causing the two to get out of sync.  Try to avoid straining the motors when choosing the final values.


### Configuring for Home Assistant
HA's MQTT is used to interact with the Blinds-Controller.  Based on the MQTT topic setting in the example above using `my_blinds` as the mid-tier naming, this configuration example has two blinds that are controlled by one Blinds-Controller. 

The topics of interest are:
- `cmnd/my_blinds/pwmN` _pwmN-value_   This is the command HA sends to the Blinds-Controller  to set the tilt position where N is the pwm channel configured for that particular blind.
- `stat/my_blinds/PWM {"PWM":{"PWMN":pwm1-value,"PWM2":pwm2-value}}`  This is the status the Blinds-Controller reports back to HA.  It reports the PWM value of both blinds in one report.
- `tele/my_blinds/LWT Online`   This is the Last Will and Testament topic sent by the Broker to HA (here it is showing as Online).

The following shows an example of a single Blinds Controller controlling two independent set of blinds.  The state of the blinds "open" or "closed" will be based on the tile angle of the slats.

For Blind#1
```
cover:
  - platform: mqtt
    name: "My Blinds Right"
    state_topic: "stat/my_blinds/PWM"
    state_open: "on"
    state_closed: "off"
    value_template: '{% if value_json["PWM"]["PWM1"] == 24 -%}off{%- else -%}on{%- endif %}'
    tilt_command_topic: "cmnd/my_blinds/pwm1"
    tilt_min: 24                #PWM min theoretically 28; Intentionally overshifted.
    tilt_max: 123               #PWM max theoreticaly 123
    tilt_closed_value: 24       #PWM value
    tilt_opened_value: 84       #PWM value. 90 degrees theoretically 73
    tilt_status_topic: "stat/my_blinds/PWM"
    tilt_status_template: '{{ value_json["PWM"]["PWM1"] }}'
   #tilt_status_optimistic: true #not needed if tilt_status_topic is used.
    availability_topic: "tele/my_blinds/LWT"
    payload_available: "Online"
    payload_not_available: "Offline"
    retain: true
   #optimistic: true
```
For Blinds#1, the PWM value 24 has the blind's slats closed; value 84 has the slots horizontal and is the position we want when HA is told to "open" the blinds; value 123 has the blind slats tilted in the opposite direction.  Home Assistant will convert these values into a percentage 0 to 100%.


For Blind#2:
```
  - platform: mqtt
    name: "My Blinds Left"
    state_topic: "stat/my_blinds/PWM"
    state_open: "on"
    state_closed: "off"
    value_template: '{% if value_json["PWM"]["PWM2"] == 28 -%}off{%- else -%}on{%- endif %}'
    tilt_command_topic: "cmnd/my_blinds/pwm2"
    tilt_min: 28                #PWM min theoretically 28; 
    tilt_max: 123               #PWM max theoreticaly 123
    tilt_closed_value: 28       #PWM value
    tilt_opened_value: 77       #PWM value. 90 degrees theoretically 73
    tilt_status_topic: "stat/my_blinds/PWM"
    tilt_status_template: '{{ value_json["PWM"]["PWM2"] }}'
   #tilt_status_optimistic: true #not needed if tilt_status_topic is used.
    availability_topic: "tele/my_blinds/LWT"
    payload_available: "Online"
    payload_not_available: "Offline"
    retain: true
    #optimistic: true

```
For Blinds#2, the PWM values (in this case PWM2) are slightly different due to the variability mentioned above between servos and their attachments to the tilt mechanisms.

And finally you can group these two independent blinds into one:
```
  - platform: group
    name: My Blinds
    entities:
      - cover.my_blinds_right
      - cover.my_blinds_left
```
As a note, there are a few quirks in Home Assistant's Group entity for tilt covers.  For example when both blinds are moved to the open position (`tilt_opened_value`), the cover Group entity will show the tilt position as 100%, even though the two blind's open position is say 54 and 61 percent respectively.

# Credits and License
Credit goes to [arendst/Tasmota](https://github.com/arendst/Tasmota), formerly
https://github.com/arendst/Sonoff-Tasmota, for the original code base.

This program is licensed under GPL-3.0


RELEASENOTES.md
## Base Code
For use with Arduino IDE.

## Release notes
### Initial Release.
Version 6.2.1.3
Code is based on Sonoff-Tasmota version 6.2.1 and ESP Core version 2.3.0.



sonoff/sonoff_version.h
/*
  sonoff_version.h - Version header file for Sonoff-Tasmota

  Copyright (C) 2018  Theo Arends

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef _SONOFF_VERSION_H_
#define _SONOFF_VERSION_H_

//#define VERSION            0x06020100
#define VERSION            0x06020103

#define D_PROGRAMNAME      "Sonoff-Tasmota-Blinds"
#define D_AUTHOR           "Theo Arends (modified by Tommy Long for Blinds)"
//#define D_WEBLINK          "https://github.com/arendst/Sonoff-Tasmota"
#define D_WEBLINK          "https://bit.ly/tasmota"

#endif  // _SONOFF_VERSION_H_


sonoff/settings.h
/*
  settings.h - setting variables for Sonoff-Tasmota

  Copyright (C) 2018  Theo Arends

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef _SETTINGS_H_
#define _SETTINGS_H_

#define PARAM8_SIZE  18                    // Number of param bytes (SetOption)

typedef union {                            // Restricted by MISRA-C Rule 18.4 but so useful...
  uint32_t data;                           // Allow bit manipulation using SetOption
  struct {                                 // SetOption0 .. SetOption31
    uint32_t save_state : 1;               // bit 0
    uint32_t button_restrict : 1;          // bit 1
    uint32_t value_units : 1;              // bit 2
    uint32_t mqtt_enabled : 1;             // bit 3
    uint32_t mqtt_response : 1;            // bit 4
    uint32_t mqtt_power_retain : 1;        // CMND_POWERRETAIN
    uint32_t mqtt_button_retain : 1;       // CMND_BUTTONRETAIN
    uint32_t mqtt_switch_retain : 1;       // CMND_SWITCHRETAIN
    uint32_t temperature_conversion : 1;   // bit 8
    uint32_t mqtt_sensor_retain : 1;       // CMND_SENSORRETAIN
    uint32_t mqtt_offline : 1;             // bit 10
    uint32_t button_swap : 1;              // bit 11 (v5.1.6)
    uint32_t stop_flash_rotate : 1;        // bit 12 (v5.2.0)
    uint32_t button_single : 1;            // bit 13 (v5.4.0)
    uint32_t interlock : 1;                // bit 14 (v5.6.0)
    uint32_t pwm_control : 1;              // bit 15 (v5.8.1)
    uint32_t ws_clock_reverse : 1;         // bit 16 (v5.8.1)
    uint32_t decimal_text : 1;             // bit 17 (v5.8.1)
    uint32_t light_signal : 1;             // bit 18 (v5.10.0c)
    uint32_t hass_discovery : 1;           // bit 19 (v5.11.1a)
    uint32_t not_power_linked : 1;         // bit 20 (v5.11.1f)
    uint32_t no_power_on_check : 1;        // bit 21 (v5.11.1i)
    uint32_t mqtt_serial : 1;              // bit 22 (v5.12.0f)
    uint32_t mqtt_serial_raw : 1;          // bit 23 (v6.1.1c)
    uint32_t rules_once : 1;               // bit 24 (v5.12.0k) - free since v5.14.0b
    uint32_t knx_enabled : 1;              // bit 25 (v5.12.0l) KNX
    uint32_t device_index_enable : 1;      // bit 26 (v5.13.1a)
    uint32_t knx_enable_enhancement : 1;   // bit 27 (v5.14.0a) KNX
    uint32_t rf_receive_decimal : 1;       // bit 28 (v6.0.0a)
    uint32_t ir_receive_decimal : 1;       // bit 29 (v6.0.0a)
    uint32_t hass_light : 1;               // bit 30 (v6.0.0b)
    uint32_t global_state : 1;             // bit 31 (v6.1.0)
  };
} SysBitfield;

typedef union {                            // Restricted by MISRA-C Rule 18.4 but so useful...
  uint32_t data;                           // Allow bit manipulation using SetOption
  struct {                                 // SetOption50 .. SetOption81
    uint32_t timers_enable : 1;            // bit 0 (v6.1.1b)
    uint32_t spare01 : 1;
    uint32_t spare02 : 1;
    uint32_t spare03 : 1;
    uint32_t spare04 : 1;
    uint32_t spare05 : 1;
    uint32_t spare06 : 1;
    uint32_t spare07 : 1;
    uint32_t spare08 : 1;
    uint32_t spare09 : 1;
    uint32_t spare10 : 1;
    uint32_t spare11 : 1;
    uint32_t spare12 : 1;
    uint32_t spare13 : 1;
    uint32_t spare14 : 1;
  //uint32_t spare15 : 1;
    uint32_t slow_servo : 1;               //TOMMY Option65 for slow servo mode
  //uint32_t spare16 : 1;
    uint32_t slow_servo_retain_results: 1; //TOMMY Option66 to retain MQTT slow servo results/status
    uint32_t spare17 : 1;
    uint32_t spare18 : 1;
    uint32_t spare19 : 1;
    uint32_t spare20 : 1;
    uint32_t spare21 : 1;
    uint32_t spare22 : 1;
    uint32_t spare23 : 1;
    uint32_t spare24 : 1;
    uint32_t spare25 : 1;
    uint32_t spare26 : 1;
    uint32_t spare27 : 1;
    uint32_t spare28 : 1;
    uint32_t spare29 : 1;
    uint32_t spare30 : 1;
    uint32_t user_esp8285_enable : 1;      // bit 31 (v6.1.1.14)
  };
} SysBitfield3;

typedef union {
  uint32_t data;                           // Allow bit manipulation
  struct {
    uint32_t spare00 : 1;
    uint32_t spare01 : 1;
    uint32_t spare02 : 1;
    uint32_t spare03 : 1;
    uint32_t spare04 : 1;
    uint32_t spare05 : 1;
    uint32_t spare06 : 1;
    uint32_t spare07 : 1;
    uint32_t spare08 : 1;
    uint32_t spare09 : 1;
    uint32_t spare10 : 1;
    uint32_t spare11 : 1;
    uint32_t spare12 : 1;
    uint32_t axis_resolution : 2;
    uint32_t current_resolution : 2;
    uint32_t voltage_resolution : 2;
    uint32_t wattage_resolution : 2;
    uint32_t emulation : 2;
    uint32_t energy_resolution : 3;
    uint32_t pressure_resolution : 2;
    uint32_t humidity_resolution : 2;
    uint32_t temperature_resolution : 2;
  };
} SysBitfield2;

typedef union {
  uint16_t data;
  struct {
    uint16_t hemis : 1;                    // bit 0        = 0=Northern, 1=Southern Hemisphere (=Opposite DST/STD)
    uint16_t week : 3;                     // bits 1 - 3   = 0=Last week of the month, 1=First, 2=Second, 3=Third, 4=Fourth
    uint16_t month : 4;                    // bits 4 - 7   = 1=Jan, 2=Feb, ... 12=Dec
    uint16_t dow : 3;                      // bits 8 - 10  = day of week, 1=Sun, 2=Mon, ... 7=Sat
    uint16_t hour : 5;                     // bits 11 - 15 = 0-23
  };
} TimeRule;

typedef union {
  uint32_t data;
  struct {
    uint32_t time : 11;                    // bits 0 - 10 = minutes in a day
    uint32_t window : 4;                   // bits 11 - 14 = minutes random window
    uint32_t repeat : 1;                   // bit 15
    uint32_t days : 7;                     // bits 16 - 22 = week day mask
    uint32_t device : 4;                   // bits 23 - 26 = 16 devices
    uint32_t power : 2;                    // bits 27 - 28 = 4 power states - Off, On, Toggle, Blink or Rule
    uint32_t mode : 2;                     // bits 29 - 30 = timer modes - 0 = Scheduler, 1 = Sunrise, 2 = Sunset
    uint32_t arm : 1;                      // bit 31
  };
} Timer;

typedef union {
  uint16_t data;
  struct {
    uint16_t pinmode : 3;                   // Pin mode (1 through 6)
    uint16_t pullup : 1;                    // Enable internal weak pull-up resistor
    uint16_t saved_state : 1;               // Save output state, if used.
    uint16_t int_report_mode : 2;           // Interrupt reporting mode 0 = immediate telemetry & event, 1 = immediate event only, 2 = immediate telemetry only
    uint16_t int_report_defer : 4;          // Number of interrupts to ignore until reporting (default 0, max 15)
    uint16_t int_count_en : 1;              // Enable interrupt counter for this pin
    uint16_t spare12 : 1;
    uint16_t spare13 : 1;
    uint16_t spare14 : 1;
    uint16_t spare15 : 1;
  };
} Mcp230xxCfg;

/*
struct SYSCFG {
  unsigned long cfg_holder;                // 000 Pre v6 header
  unsigned long save_flag;                 // 004
  unsigned long version;                   // 008
  unsigned long bootcount;                 // 00C
*/
struct SYSCFG {
  uint16_t cfg_holder;                     // 000 v6 header
  uint16_t cfg_size;                       // 002
  unsigned long save_flag;                 // 004
  unsigned long version;                   // 008
  uint16_t bootcount;                      // 00C
  uint16_t cfg_crc;                        // 00E
  SysBitfield   flag;                      // 010
  int16_t       save_data;                 // 014
  int8_t        timezone;                  // 016
  char          ota_url[101];              // 017
  char          mqtt_prefix[3][11];        // 07C
  uint8_t       baudrate;                  // 09D
  byte          seriallog_level;           // 09E
  uint8_t       sta_config;                // 09F
  byte          sta_active;                // 0A0
  char          sta_ssid[2][33];           // 0A1
  char          sta_pwd[2][65];            // 0E3
  char          hostname[33];              // 165
  char          syslog_host[33];           // 186
  uint8_t       rule_stop;                 // 1A7
  uint16_t      syslog_port;               // 1A8
  byte          syslog_level;              // 1AA
  uint8_t       webserver;                 // 1AB
  byte          weblog_level;              // 1AC
  uint8_t       mqtt_fingerprint[2][20];   // 1AD

  byte          free_1D5[20];              // 1D5  Free since 5.12.0e

  char          mqtt_host[33];             // 1E9
  uint16_t      mqtt_port;                 // 20A
  char          mqtt_client[33];           // 20C
  char          mqtt_user[33];             // 22D
  char          mqtt_pwd[33];              // 24E
  char          mqtt_topic[33];            // 26F
  char          button_topic[33];          // 290
  char          mqtt_grptopic[33];         // 2B1
  uint8_t       display_model;             // 2D2
  uint8_t       display_mode;              // 2D3
  uint8_t       display_refresh;           // 2D4
  uint8_t       display_rows;              // 2D5
  uint8_t       display_cols[2];           // 2D6
  uint8_t       display_address[8];        // 2D8
  uint8_t       display_dimmer;            // 2E0
  uint8_t       display_size;              // 2E1
  TimeRule      tflag[2];                  // 2E2
  uint16_t      pwm_frequency;             // 2E6
  power_t       power;                     // 2E8
  uint16_t      pwm_value[MAX_PWMS];       // 2EC
  int16_t       altitude;                  // 2F6
  uint16_t      tele_period;               // 2F8
  uint8_t       display_rotate;            // 2FA
  uint8_t       ledstate;                  // 2FB
  uint8_t       param[PARAM8_SIZE];        // 2FC  SetOption32 .. SetOption49
  int16_t       toffset[2];                // 30E
  uint8_t       display_font;              // 312
  char          state_text[4][11];         // 313
  uint8_t       energy_power_delta;        // 33F
  uint16_t      domoticz_update_timer;     // 340
  uint16_t      pwm_range;                 // 342
  unsigned long domoticz_relay_idx[MAX_DOMOTICZ_IDX];  // 344
  unsigned long domoticz_key_idx[MAX_DOMOTICZ_IDX];    // 354
  unsigned long energy_power_calibration;  // 364
  unsigned long energy_voltage_calibration;  // 368
  unsigned long energy_current_calibration;  // 36C
  unsigned long energy_kWhtoday;           // 370
  unsigned long energy_kWhyesterday;       // 374
  uint16_t      energy_kWhdoy;             // 378
  uint16_t      energy_min_power;          // 37A
  uint16_t      energy_max_power;          // 37C
  uint16_t      energy_min_voltage;        // 37E
  uint16_t      energy_max_voltage;        // 380
  uint16_t      energy_min_current;        // 382
  uint16_t      energy_max_current;        // 384
  uint16_t      energy_max_power_limit;    // 386 MaxPowerLimit
  uint16_t      energy_max_power_limit_hold;         // 388 MaxPowerLimitHold
  uint16_t      energy_max_power_limit_window;       // 38A MaxPowerLimitWindow
  uint16_t      energy_max_power_safe_limit;         // 38C MaxSafePowerLimit
  uint16_t      energy_max_power_safe_limit_hold;    // 38E MaxSafePowerLimitHold
  uint16_t      energy_max_power_safe_limit_window;  // 390 MaxSafePowerLimitWindow
  uint16_t      energy_max_energy;         // 392 MaxEnergy
  uint16_t      energy_max_energy_start;   // 394 MaxEnergyStart
  uint16_t      mqtt_retry;                // 396
  uint8_t       poweronstate;              // 398
  uint8_t       last_module;               // 399
  uint16_t      blinktime;                 // 39A
  uint16_t      blinkcount;                // 39C
  uint16_t      light_rotation;            // 39E
  SysBitfield3  flag3;                     // 3A0
  uint8_t       switchmode[MAX_SWITCHES];  // 3A4  (6.0.0b - moved from 0x4CA)
  char          friendlyname[MAX_FRIENDLYNAMES][33]; // 3AC
  char          switch_topic[33];          // 430
  char          serial_delimiter;          // 451
  uint8_t       sbaudrate;                 // 452
  uint8_t       sleep;                     // 453
  uint16_t      domoticz_switch_idx[MAX_DOMOTICZ_IDX];      // 454
  uint16_t      domoticz_sensor_idx[MAX_DOMOTICZ_SNS_IDX];  // 45C
  uint8_t       module;                    // 474
  uint8_t       ws_color[4][3];            // 475
  uint8_t       ws_width[3];               // 481
  myio          my_gp;                     // 484
  uint16_t      light_pixels;              // 496
  uint8_t       light_color[5];            // 498
  uint8_t       light_correction;          // 49D
  uint8_t       light_dimmer;              // 49E
  uint8_t       rule_enabled;              // 49F
  uint8_t       rule_once;                 // 4A0
  uint8_t       light_fade;                // 4A1
  uint8_t       light_speed;               // 4A2
  uint8_t       light_scheme;              // 4A3
  uint8_t       light_width;               // 4A4
  byte          knx_GA_registered;         // 4A5  Number of Group Address to read
  uint16_t      light_wakeup;              // 4A6
  byte          knx_CB_registered;         // 4A8  Number of Group Address to write
  char          web_password[33];          // 4A9

  uint8_t       ex_switchmode[4];          // 4CA  Free since 6.0.0a

  char          ntp_server[3][33];         // 4CE
  byte          ina219_mode;               // 531
  uint16_t      pulse_timer[MAX_PULSETIMERS]; // 532
  uint16_t      button_debounce;           // 542
  uint32_t      ip_address[4];             // 544
  unsigned long energy_kWhtotal;           // 554
  char          mqtt_fulltopic[100];       // 558
  SysBitfield2  flag2;                     // 5BC
  unsigned long pulse_counter[MAX_COUNTERS];  // 5C0
  uint16_t      pulse_counter_type;        // 5D0
  uint16_t      pulse_counter_debounce;    // 5D2
  uint8_t       rf_code[17][9];            // 5D4

  byte          free_66d[1];               // 66D

  uint16_t      switch_debounce;           // 66E
  Timer         timer[MAX_TIMERS];         // 670
  int           latitude;                  // 6B0
  int           longitude;                 // 6B4
  uint16_t      knx_physsical_addr;        // 6B8  (address_t is a uint16_t)
  uint16_t      knx_GA_addr[MAX_KNX_GA];   // 6BA  (address_t is a uint16_t) x KNX_max_GA
  uint16_t      knx_CB_addr[MAX_KNX_CB];   // 6CE  (address_t is a uint16_t) x KNX_max_CB
  byte          knx_GA_param[MAX_KNX_GA];  // 6E2  Type of Input (relay changed, button pressed, sensor read <-teleperiod)
  byte          knx_CB_param[MAX_KNX_CB];  // 6EC  Type of Output (set relay, toggle relay, reply sensor value)
  Mcp230xxCfg   mcp230xx_config[16];       // 6F6
  uint8_t       mcp230xx_int_prio;         // 716

  byte          free_717[1];               // 717

  uint16_t      mcp230xx_int_timer;        // 718

  byte          free_71A[180];             // 71A

  char          mems[MAX_RULE_MEMS][10];  // 7CE
                                           // 800 Full - no more free locations

  char          rules[MAX_RULE_SETS][MAX_RULE_SIZE]; // 800 uses 512 bytes in v5.12.0m, 3 x 512 bytes in v5.14.0b

                                           // E00 - FFF free locations
} Settings;

struct RTCRBT {
  uint16_t      valid;                     // 000
  uint8_t       fast_reboot_count;         // 002
  uint8_t       free_003[1];               // 003
} RtcReboot;

struct RTCMEM {
  uint16_t      valid;                     // 000
  byte          oswatch_blocked_loop;      // 002
  uint8_t       ota_loader;                // 003
  unsigned long energy_kWhtoday;              // 004
  unsigned long energy_kWhtotal;              // 008
  unsigned long pulse_counter[MAX_COUNTERS];  // 00C
  power_t       power;                     // 01C
  uint8_t       free_020[60];              // 020
                                           // 05C next free location (64 (=core) + 100 (=tasmota offset) + 92 (=0x5C RTCMEM struct) = 256 bytes (max = 512))
} RtcSettings;

struct TIME_T {
  uint8_t       second;
  uint8_t       minute;
  uint8_t       hour;
  uint8_t       day_of_week;               // sunday is day 1
  uint8_t       day_of_month;
  uint8_t       month;
  char          name_of_month[4];
  uint16_t      day_of_year;
  uint16_t      year;
  unsigned long days;
  unsigned long valid;
} RtcTime;

struct XDRVMAILBOX {
  uint16_t      valid;
  uint16_t      index;
  uint16_t      data_len;
  uint16_t      payload16;
  int16_t       payload;
  uint8_t       grpflg;
  uint8_t       notused;
  char         *topic;
  char         *data;
} XdrvMailbox;

#define MAX_RULES_FLAG  7                  // Number of bits used in RulesBitfield (tricky I know...)
typedef union {                            // Restricted by MISRA-C Rule 18.4 but so useful...
  uint16_t data;                           // Allow bit manipulation
  struct {
    uint16_t system_boot : 1;
    uint16_t time_init : 1;
    uint16_t time_set : 1;
    uint16_t mqtt_connected : 1;
    uint16_t mqtt_disconnected : 1;
    uint16_t wifi_connected : 1;
    uint16_t wifi_disconnected : 1;
    uint16_t spare07 : 1;
    uint16_t spare08 : 1;
    uint16_t spare09 : 1;
    uint16_t spare10 : 1;
    uint16_t spare11 : 1;
    uint16_t spare12 : 1;
    uint16_t spare13 : 1;
    uint16_t spare14 : 1;
    uint16_t spare15 : 1;
  };
} RulesBitfield;

typedef union {
  uint8_t data;
  struct {
    uint8_t wifi_down : 1;
    uint8_t mqtt_down : 1;
    uint8_t spare02 : 1;
    uint8_t spare03 : 1;
    uint8_t spare04 : 1;
    uint8_t spare05 : 1;
    uint8_t spare06 : 1;
    uint8_t spare07 : 1;
  };
} StateBitfield;

// See issue https://github.com/esp8266/Arduino/issues/2913
#ifdef USE_ADC_VCC
  ADC_MODE(ADC_VCC);                       // Set ADC input for Power Supply Voltage usage
#endif

#endif  // _SETTINGS_H_
